src/wallet/wallet2.cpp

```diff
@@ -13055,6 +13065,75 @@ void wallet2::throw_on_rpc_response_error(const boost::optional<std::string> &st
   THROW_WALLET_EXCEPTION_IF(*status != CORE_RPC_STATUS_OK, tools::error::wallet_generic_rpc_error, method, m_trusted_daemon ? *status : "daemon error");
 }
 //----------------------------------------------------------------------------------------------------
+
+bool wallet2::save_to_file(const std::string& path_to_file, const std::string& raw, bool is_printable) const
+{
+  if (is_printable || m_export_format == ExportFormat::Binary)
+  {
+    return epee::file_io_utils::save_string_to_file(path_to_file, raw);
+  }
+
+  FILE *fp = fopen(path_to_file.c_str(), "w+");
+  // Save the result b/c we need to close the fp before returning success/failure.
+  int write_result = PEM_write(fp, ASCII_OUTPUT_MAGIC.c_str(), "", (const unsigned char *) raw.c_str(), raw.length());
+  fclose(fp);
+
+  if (write_result == 0)
+  {
+    return false;
+  }
+  else
+  {
+    return true;
+  }
+}
+//----------------------------------------------------------------------------------------------------
+
+bool wallet2::load_from_file(const std::string& path_to_file, std::string& target_str,
+                             size_t max_size)
+{
+  std::string data;
+  bool r = epee::file_io_utils::load_file_to_string(path_to_file, data, max_size);
+  if (!r)
+  {
+    return false;
+  }
+
+  if (!boost::algorithm::contains(boost::make_iterator_range(data.begin(), data.end()), ASCII_OUTPUT_MAGIC))
+  {
+    // It's NOT our ascii dump.
+    target_str = std::move(data);
+    return true;
+  }
+
+  // Creating a BIO and calling PEM_read_bio instead of simpler PEM_read
+  // to avoid reading the file from disk twice.
+  BIO* b = BIO_new_mem_buf((const void*) data.data(), data.length());
+
+  char *name = NULL;
+  char *header = NULL;
+  unsigned char *openssl_data = NULL;
+  long len = 0;
+
+  // Save the result b/c we need to free the data before returning success/failure.
+  int read_result = PEM_read_bio(b, &name, &header, &openssl_data, &len);
+
+  target_str = std::string((const char*) openssl_data, len);
+  OPENSSL_free((void *) name);
```

Comment:
`name`, `header`, and `openssl_data` leak if `std::string` constructor throws. Unfortunately I do not see a C++ functor for use with `unique_ptr` in our code base. Its not difficult to add one. Add a todo if you don't have time to add and test now as the case is not likely to trigger.

Reply:
Good catch about the exception. I can either:

1. Just slap on the todo.
2. Simply wrap it in a try/catch block
3. Add the suggested functor. I'm not sure I 100% understood what you mean - something like this? Where do you think I should put such a helper function?

```c++
template <typename... Targs>
std::unique_ptr<std::string> safe_string(Targs... args)
{
    try
    {                                                                                                                                                                                                                                                                                   return std::make_unique<std::string>(std::string(args...));
    }
    catch (...)
    {
        return std::make_unique<std::string>(nullptr);
    }
}
```

Reply:
No I meant add the `unique_ptr` for the `OPENSSL_free` calls:
```c++
struct openssl_free
{
    void operator()(void* ptr) const noexcept
    {
        if (ptr)
            OPENSSL_free(ptr);
     }
};
...
std::unique_ptr<char, openssl_free> name;
std::unique_ptr<char, openssl_free> header;
std::unique_ptr<unsigned char, openssl_free> openssl_data;
...
```
The annoying part are the out parameters, because it requires a temporary stack value that gets copied into the `std::unique_ptr` before doing any throwing operation. Having a `try { } catch (...) {}` around the construction is probably the easiest. The solution you had would work too, although I'm not sure how often the function would be used.

Reply:
I see, thanks. I'd go with the easy solution of try/catch block around the constructor because the other solutions have their own drawback - either needlessly complex and generic for a single use purpose, or the temporary stack values.